File: internal/services/movie_service.go
package services

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"sort"
	"strings"

	"github.com/mbourgeois-fetch/sample-qdev-movies/internal/models"
)

// MovieService handles all movie-related business logic
type MovieService struct {
	movies []models.Movie
}

// NewMovieService creates a new instance of MovieService
func NewMovieService() *MovieService {
	service := &MovieService{}
	service.loadMovies()
	return service
}

// loadMovies loads movie data from JSON file
func (s *MovieService) loadMovies() {
	log.Println("🏴‍☠️ Loading movie treasure from movies.json...")

	data, err := ioutil.ReadFile("data/movies.json")
	if err != nil {
		log.Printf("Arrr! Error reading movies.json: %v", err)
		return
	}

	err = json.Unmarshal(data, &s.movies)
	if err != nil {
		log.Printf("Shiver me timbers! Error parsing movies.json: %v", err)
		return
	}

	log.Printf("⚓ Successfully loaded %d movies into our treasure chest!", len(s.movies))
}

// GetAllMovies returns all movies in our treasure chest
func (s *MovieService) GetAllMovies() []models.Movie {
	return s.movies
}

// GetMovieByID finds a specific movie by its ID
func (s *MovieService) GetMovieByID(id int64) (*models.Movie, error) {
	if id <= 0 {
		return nil, fmt.Errorf("invalid movie ID: %d", id)
	}

	for _, movie := range s.movies {
		if movie.ID == id {
			return &movie, nil
		}
	}

	return nil, fmt.Errorf("movie with ID %d not found", id)
}

// SearchMovies searches through our treasure chest of movies like a true pirate!
// This method filters movies based on the search criteria ye provide, matey.
func (s *MovieService) SearchMovies(movieName string, movieID *int64, genre string) []models.Movie {
	log.Printf("🔍 Ahoy! Searching for movies with criteria - Name: %s, ID: %v, Genre: %s", 
		movieName, movieID, genre)

	var results []models.Movie

	for _, movie := range s.movies {
		if s.matchesSearchCriteria(movie, movieName, movieID, genre) {
			results = append(results, movie)
		}
	}

	log.Printf("⚓ Found %d movies matching the search criteria, ye savvy pirate!", len(results))
	return results
}

// matchesSearchCriteria checks if a movie matches our search criteria
func (s *MovieService) matchesSearchCriteria(movie models.Movie, movieName string, movieID *int64, genre string) bool {
	// If searching by ID, that takes precedence over other criteria
	if movieID != nil && *movieID > 0 {
		return movie.ID == *movieID
	}

	nameMatches := true
	genreMatches := true

	// Check movie name (case-insensitive partial match)
	if movieName != "" && strings.TrimSpace(movieName) != "" {
		nameMatches = strings.Contains(
			strings.ToLower(movie.MovieName),
			strings.ToLower(strings.TrimSpace(movieName)),
		)
	}

	// Check genre (case-insensitive partial match)
	if genre != "" && strings.TrimSpace(genre) != "" {
		genreMatches = strings.Contains(
			strings.ToLower(movie.Genre),
			strings.ToLower(strings.TrimSpace(genre)),
		)
	}

	return nameMatches && genreMatches
}

// GetAllGenres returns all unique genres from our movie treasure chest
// This be useful for showing available genres to search through
func (s *MovieService) GetAllGenres() []string {
	genreMap := make(map[string]bool)
	
	// Collect unique genres
	for _, movie := range s.movies {
		// Handle genres that might have multiple values separated by "/"
		genres := strings.Split(movie.Genre, "/")
		for _, g := range genres {
			trimmed := strings.TrimSpace(g)
			if trimmed != "" {
				genreMap[trimmed] = true
			}
		}
	}

	// Convert to sorted slice
	var genres []string
	for genre := range genreMap {
		genres = append(genres, genre)
	}
	sort.Strings(genres)
	return genres
}

File: src/main/java/com/amazonaws/samples/qdevmovies/movies/MovieService.java
package com.amazonaws.samples.qdevmovies.movies;

import org.json.JSONArray;
import org.json.JSONObject;
import org.springframework.stereotype.Service;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Scanner;
import java.util.stream.Collectors;

@Service
public class MovieService {
    private static final Logger logger = LogManager.getLogger(MovieService.class);
    private final List<Movie> movies;
    private final Map<Long, Movie> movieMap;

    public MovieService() {
        this.movies = loadMoviesFromJson();
        this.movieMap = new HashMap<>();
        for (Movie movie : movies) {
            movieMap.put(movie.getId(), movie);
        }
    }

    private List<Movie> loadMoviesFromJson() {
        List<Movie> movieList = new ArrayList<>();
        try {
            InputStream inputStream = getClass().getClassLoader().getResourceAsStream("movies.json");
            if (inputStream != null) {
                Scanner scanner = new Scanner(inputStream, StandardCharsets.UTF_8.name());
                String jsonContent = scanner.useDelimiter("\\A").next();
                scanner.close();
                
                JSONArray moviesArray = new JSONArray(jsonContent);
                for (int i = 0; i < moviesArray.length(); i++) {
                    JSONObject movieObj = moviesArray.getJSONObject(i);
                    movieList.add(new Movie(
                        movieObj.getLong("id"),
                        movieObj.getString("movieName"),
                        movieObj.getString("director"),
                        movieObj.getInt("year"),
                        movieObj.getString("genre"),
                        movieObj.getString("description"),
                        movieObj.getInt("duration"),
                        movieObj.getDouble("imdbRating")
                    ));
                }
            }
        } catch (Exception e) {
            logger.error("Failed to load movies from JSON: {}", e.getMessage());
        }
        return movieList;
    }

    public List<Movie> getAllMovies() {
        return movies;
    }

    public Optional<Movie> getMovieById(Long id) {
        if (id == null || id <= 0) {
            return Optional.empty();
        }
        return Optional.ofNullable(movieMap.get(id));
    }

    /**
     * Ahoy! Search through our treasure chest of movies like a true pirate!
     * This method filters movies based on the search criteria ye provide, matey.
     * 
     * @param movieName The name of the movie to search for (case-insensitive, partial matches allowed)
     * @param movieId The specific ID of a movie ye be seekin'
     * @param genre The genre to filter by (case-insensitive, partial matches allowed)
     * @return A list of movies that match yer search criteria, or an empty list if no treasure be found
     */
    public List<Movie> searchMovies(String movieName, Long movieId, String genre) {
        logger.info("Ahoy! Searching for movies with criteria - Name: {}, ID: {}, Genre: {}", 
                   movieName, movieId, genre);
        
        List<Movie> searchResults = movies.stream()
            .filter(movie -> matchesSearchCriteria(movie, movieName, movieId, genre))
            .collect(Collectors.toList());
            
        logger.info("Found {} movies matching the search criteria, ye savvy pirate!", searchResults.size());
        return searchResults;
    }

    /**
     * Arrr! This be the crew member that checks if a movie matches our search criteria.
     * 
     * @param movie The movie to examine
     * @param movieName Name to search for (can be null or empty)
     * @param movieId ID to search for (can be null)
     * @param genre Genre to search for (can be null or empty)
     * @return true if the movie matches all provided criteria, false otherwise
     */
    private boolean matchesSearchCriteria(Movie movie, String movieName, Long movieId, String genre) {
        // If searching by ID, that takes precedence over other criteria
        if (movieId != null && movieId > 0) {
            return movie.getId() == movieId;
        }
        
        boolean nameMatches = true;
        boolean genreMatches = true;
        
        // Check movie name (case-insensitive partial match)
        if (movieName != null && !movieName.trim().isEmpty()) {
            nameMatches = movie.getMovieName().toLowerCase()
                              .contains(movieName.trim().toLowerCase());
        }
        
        // Check genre (case-insensitive partial match)
        if (genre != null && !genre.trim().isEmpty()) {
            genreMatches = movie.getGenre().toLowerCase()
                               .contains(genre.trim().toLowerCase());
        }
        
        return nameMatches && genreMatches;
    }

    /**
     * Batten down the hatches! Get all unique genres from our movie treasure chest.
     * This be useful for showing available genres to search through.
     * 
     * @return A list of all unique genres in our movie collection
     */
    public List<String> getAllGenres() {
        return movies.stream()
            .map(Movie::getGenre)
            .distinct()
            .sorted()
            .collect(Collectors.toList());
    }
}
